### PR TITLE
Starting to replace Xpress. by Lib. in MOI

### PR DESCRIPTION
**Pull request by RaphMota**

### CODE REVIEW

File: src/MOI/MOI_callbacks.jl [hunk 1]
```diff
@@ -59,12 +62,15 @@ end
...
+    
     return (cb_data) -> begin
         get_cb_solution(model, cb_data.model)
         if model.heuristic_callback !== nothing
             model.callback_state = CB_HEURISTIC
             # only allow one heuristic solution per LP optimal node
-            if Xpress.getintattrib(cb_data.model, Xpress.Lib.XPRS_CALLBACKCOUNT_OPTNODE) > 1
+
+            if Xpress.getintattrib(cb_data.model, Lib.XPRS_CALLBACKCOUNT_OPTNODE) > 1
+                #println("Attribute ",Lib.XPRS_CALLBACKCOUNT_OPTNODE," value recovered : ",Lib.XPRSgetintattrib(cb_data.model, Lib.XPRS_CALLBACKCOUNT_OPTNODE, _))
```
Comment (guilhermebodin) [inline src/MOI/MOI_callbacks.jl:73]: These were not here before, could you remove them?

File: src/MOI/MOI_callbacks.jl [hunk 2]
```diff
@@ -81,7 +87,8 @@ function default_moi_callback(model::Optimizer)
             # only allow one user cut solution per LP optimal node
             # limiting two calls to guarantee th user has a chance to add
             # a cut. if the user cut is loose the problem will be resolved anyway.
-            if Xpress.getintattrib(cb_data.model, Xpress.Lib.XPRS_CALLBACKCOUNT_OPTNODE) > 2
+            if Xpress.getintattrib(cb_data.model, Lib.XPRS_CALLBACKCOUNT_OPTNODE)> 2
+                #println("Attribute ",Lib.XPRS_CALLBACKCOUNT_OPTNODE," value recovered : ",Lib.XPRSgetintattrib(cb_data.model, Lib.XPRS_CALLBACKCOUNT_OPTNODE, _)::Int)
```
Comment (guilhermebodin) [inline src/MOI/MOI_callbacks.jl:91]: same thing, kind of weird, there are a lot in the code

File: src/MOI/MOI_callbacks.jl [hunk 3]
```diff
@@ -155,22 +162,26 @@ function MOI.submit(
...
-        Xpress.interrupt(model_cb, Xpress.Lib.XPRS_STOP_USER)
+            Lib.XPRSinterrupt(model_cb, Lib.XPRS_STOP_USER)
+            #println("Optimizer algorithm interrupted")
         return
     elseif model.callback_state == CB_LAZY && CB <: MOI.UserCut{CallbackData}
         cache_exception(model,
             MOI.InvalidCallbackUsage(MOI.LazyConstraintCallback(), cb))
-        Xpress.interrupt(model_cb, Xpress.Lib.XPRS_STOP_USER)
+            Lib.XPRSinterrupt(model_cb, Lib.XPRS_STOP_USER)
+            #println("Optimizer algorithm interrupted")
```
Comment (guilhermebodin) [inline src/MOI/MOI_callbacks.jl:172]: Remove all of these commented `println`

File: src/MOI/MOI_wrapper.jl [hunk 4]
```diff
@@ -323,7 +325,11 @@ function MOI.empty!(model::Optimizer)
     log_level = model.log_level
     log_level != 0 && MOI.set(model, MOI.RawOptimizerAttribute("OUTPUTLOG"), 0)
     # silently load a empty model - to avoid useless printing
-    Xpress.loadlp(model.inner)
+    
+ 
+    Lib.XPRSloadlp(model.inner,"", 0, 0, Cchar[], Float64[], Float64[], Float64[], Int[], Int[], Int[], Float64[], Float64[],Float64[])
+    #println("Result of loadlp : ",Lib.XPRSloadlp(model.inner,"", 0, 0, Cchar[], Float64[], Float64[], Float64[], Int[], Int[], Int[], Float64[], Float64[],Float64[]),". ")
```
Comment (guilhermebodin) [inline src/MOI/MOI_wrapper.jl:331]: Remove all trailing comments

File: src/MOI/MOI_wrapper.jl [hunk 5]
```diff
@@ -805,24 +814,30 @@ function MOI.add_variable(model::Optimizer)
...
-    Xpress.addcols(
+    Lib.XPRSaddcols(
         model.inner,
+        1,#length(_dbdl)::Int,
+        0,#length(_dmatval)::Int,
         [0.0],#_dobj::Vector{Float64},
-        Int[],#_mstart::Vector{Int},
-        Int[],#_mrwind::Vector{Int},
+        Cint.(Int[] .- 1),#Cint.(_mrwind::Vector{Int}),
+        Cint.(Int[] .- 1),#Cint.(_mrstart::Vector{Int}),
```
Comment (guilhermebodin) [inline src/MOI/MOI_wrapper.jl:823]: Here you could simply put `Cint[]`

File: src/MOI/MOI_wrapper.jl [hunk 6]
```diff
@@ -805,24 +814,30 @@ function MOI.add_variable(model::Optimizer)
...
-    Xpress.addcols(
+    Lib.XPRSaddcols(
         model.inner,
+        N,#length(_dbdl)::Int,
+        0,#length(_dmatval)::Int,
         zeros(N),# _dobj::Vector{Float64},
-        Int[],# _mstart::Vector{Int},
-        Int[],# _mrwind::Vector{Int},
+        Cint.(Int[] .- 1),#Cint.(_mrwind::Vector{Int}),
+        Cint.(Int[] .- 1),#Cint.(_mrstart::Vector{Int}),
```
Comment (guilhermebodin) [inline src/MOI/MOI_wrapper.jl:840]: Only `Cint[]` should be enough

File: src/MOI/testing.jl [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+@assert length(_dbdl) == length(_dbdu)
+    fixinfinity!(_dbdl)
+    fixinfinity!(_dbdu)
+    ncols = length(_dbdl)
+    ncoeffs = length(_dmatval)
+    _mstart = _mstart .- 1
+    _mrwind = _mrwind .- 1
+    XPRSaddcols(model.inner, ncols, ncoeffs, [0.0], Cint.(_mstart), Cint.(_mrwind), _dmatval, _dbdl, _dbdu)
+
+    print(Xpress.getintattrib(cb_data.model, Lib.XPRS_CALLBACKCOUNT_OPTNODE)==Lib.XPRSgetintattrib(cb_data.model, Lib.XPRS_CALLBACKCOUNT_OPTNODE, 0))
```
Comment (guilhermebodin) [inline src/MOI/testing.jl:20]: Remove the file

File: src/lib.jl [hunk 8]
```diff
@@ -2,6 +2,11 @@
 # Automatically generated using Clang.jl
 
 
+
+function addition(x,y)
+    return x+y
+end
+
```
Comment (guilhermebodin) [inline src/lib.jl:9]: Remove

File: test/runtests.jl [hunk 9]
```diff
@@ -10,7 +10,9 @@ println("Optimizer version: $(Xpress.getversion())")
     "Derivative",
 ]
     @testset "$(file)" for file in readdir(folder)
+        println("Testing ", joinpath(folder, file))
```
Comment (guilhermebodin) [inline test/runtests.jl:13]: Remove this and the trailing line

File: test/runtests.jl [hunk 10]
```diff
@@ -10,7 +10,9 @@ println("Optimizer version: $(Xpress.getversion())")
...
         include(joinpath(folder, file))
+
```
Comment (guilhermebodin) [inline test/runtests.jl:15]: Remove extra line

**Review by guilhermebodin — CHANGES_REQUESTED** [review]

File: src/MOI/MOI_callbacks.jl (hunk 1 above)
Comment (RaphMota) [reply src/MOI/MOI_callbacks.jl:73]: I see only one of them in my code, weird

File: src/MOI/MOI_callbacks.jl (hunk 2 above)
Comment (RaphMota) [reply src/MOI/MOI_callbacks.jl:91]: Same as above

File: src/MOI/MOI_callbacks.jl [hunk 11]
```diff
@@ -7,17 +7,20 @@ Set a generic Xpress callback function.
...
 end
 
 function MOI.set(model::Optimizer, ::CallbackFunction, f::Function)
+
     if model.callback_data !== nothing
-        Xpress.removecboptnode(model.inner, C_NULL, C_NULL)
+        Lib.XPRSremovecboptnode(model.inner, C_NULL, C_NULL)
+        
         model.callback_data = nothing
     end
```
Comment (guilhermebodin) [inline src/MOI/MOI_callbacks.jl:25]: @RaphMota in the changes there are a lot of new lines added that do not follow the pattern that was there previously.

**Review by guilhermebodin — CHANGES_REQUESTED** [review]

File: src/MOI/MOI_callbacks.jl (hunk 11 above)
Comment (RaphMota) [reply src/MOI/MOI_callbacks.jl]: I just removed the new lines, but there are some functions as getobj or getqrowqmatrixtriplets for which I had to add some lines in order to make them work. Do i have to remove them too?

Comment (joaquimg) [reply src/MOI/MOI_callbacks.jl]: If the lines are necessary they have to stay.
@guilhermebodin is talking about the new blank lines that are not necessary.

File: Project.toml [hunk 12]
```diff
@@ -9,6 +9,7 @@ Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
 SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
+XpressPSR = "3a57e2e4-7268-4354-91dd-d57606666fa4"
```
Comment (guilhermebodin) [inline Project.toml:12]: @RaphMota you have to remove this

**Pull request merged**